pep-0394.txt

```diff
@@ -5,104 +5,181 @@ Last-Modified: $Date$
+  means) is the recommended workaround for this problem.
+* Applications designed exclusively for a specific environment (such as
+  a container or virtual environment) may continue to use the ``python``
+  command name.
 
 These recommendations are the outcome of the relevant python-dev discussions
 in March and July 2011 ([1]_, [2]_), February 2012 ([4]_),
-September 2014 ([6]_), and discussion on GitHub in April 2018 ([7]_).
+September 2014 ([6]_), discussion on GitHub in April 2018 ([7]_),
+on python-dev in February 2019 ([8]_), and during the PEP update review
+in May 2019 ([10]_).
 
 
-Rationale
-=========
+History of this PEP
+===================
 
-This recommendation is needed as, even though the majority of distributions
-still alias the ``python`` command to Python 2, some now alias it to
+In 2011, the majority of distributions
+aliased the ``python`` command to Python 2, but some started switching it to
 Python 3 ([5]_). As some of the former distributions did not provide a
 ``python2`` command by default, there was previously no way for Python 2 code
 (or any code that invokes the Python 2 interpreter directly rather than via
 ``sys.executable``) to reliably run on all Unix-like systems without
 modification, as the ``python`` command would invoke the wrong interpreter
 version on some systems, and the ``python2`` command would fail completely
-on others. The recommendations in this PEP provide a very simple mechanism
+on others. This PEP originally provided a very simple mechanism
 to restore cross-platform support, with minimal additional work required
-on the part of distribution maintainers.
+on the part of distribution maintainers. Simplified, the recommendation was:
+
+1. The ``python`` command was preferred for code compatible with both
+   Python 2 and 3 (since it was available on all systems, even those that
+   already aliased it to Python 3).
+2. The ``python`` command should always invoke Python 2 (to prevent
+   hard-to-diagnose errors when Python 2 code is run on Python 3).
+3. The ``python2`` and ``python3`` commands should be available to specify
+   the version explicitly.
+
+However, these recommendations implicitly assumed that Python 2 would always be
+available. As Python 2 is nearing its end of life in 2020 (PEP 373, PEP 404),
+distributions are making Python 2 optional or removing it entirely.
+This means either removing the ``python`` command or switching it to invoke
+Python 3. Some distributors also decided that their users were better served by
+ignoring the PEP's recommendations, and provided system administrators with the
+freedom to configure their systems based on the needs of their particular
+environment.
+
+
+.. _rationale:
+
+Current Rationale
+=================
+
+As of 2019, activating a Python virtual environment (or its functional
+equivalent) is the best way to obtain a consistent cross-platform and
+cross-distribution experience.
```

Comment:
If you're not writing Linux distro packages, you should absolutely be using a venv for everything.

You do need to use `pipx`, and you do need to reinstall them if you change Python versions.

Reply:
If you *don't* assume the use of a venv, then everything is awful:

* `python` probably doesn't work on Windows (or runs something random)
* `py` *only* works on Windows
* `python` is only reliably available inside a venv
* `python2` isn't reliably available anywhere
* `python3` doesn't work on Windows

Reply:
Or, you run them using an explicit interpreter invocation, and forget about relying on shebang lines.

Reply:
This all feels like it's outside the scope of this PEP.  With this view, we have to endorse a third party tool (`pipx`) and a deployment workflow, and teach people how to use virtual environments (and do they use `virtualenv` or `$python -m venv`?).  The reinstallation issue is problematic too.  I don't know what any of this has to do with our recommendations to *nix distributions.  If we really want to make these recommendations, we should do that in a separate PEP, IMHO.

Reply:
One other thought: If you're using zip applications (pex, shiv), you should absolutely *not* be using venvs or `/usr/bin/env`.

Reply:
It's a response to the original draft of this PR, which advised *all* script publishers to unconditionally use `/usr/bin/python3` in their scripts, which I consider to be a bad default recommendation, since it breaks virtual environments.

`/usr/bin/env python` is better generic advice, since it respects virtual environments, and leaves the choice of specifying something else in their shebang line to folks that specifically want to support running outside a venv, and should have a better idea of which kinds of environments they want to target, and hence which compatibility trade-offs they can make in choosing an exact spelling.

Tools like `pipx`, `zipapp`, `shiv` etc all run an installer at some point, so they allow publishers to rely on console script definitions which write the shebang line at install time, rather than needing to rely on a hardcoded shebang line in the source file(s).

The relevance of all this to PEP 394 is that platform providers need to account for publisher behaviour in designing their Python user experience, and that's easier for them to do if our advice to publishers is "assume platforms will either rely on a Python installer, so you can use generated console scripts, or else that they will make themselves look somewhat like a virtual environment".